cmd/ci-operator/main.go

```diff
@@ -1380,6 +1385,64 @@ func printExecutionOrder(nodes []*api.StepNode) error {
 	return nil
 }
 
+type stepWithDependencies struct {
+	StepName     string
+	Dependencies []string
+}
+
+func dumpGraph(nodes []*api.StepNode) error {
+	artifactsEnv := os.Getenv(artifactsEnv)
+	// No target to dump to, so lets just skip this
+	if artifactsEnv == "" {
+		return nil
+	}
+
+	var result []stepWithDependencies
+	iterateAllEdges(nodes, sets.String{}, func(n *api.StepNode) {
+		r := stepWithDependencies{StepName: n.Step.Name()}
+		for _, requiment := range n.Step.Requires() {
+			iterateAllEdges(nodes, sets.String{}, func(inner *api.StepNode) {
+				if satisfiedBy(requiment, inner.Step) {
+					r.Dependencies = append(r.Dependencies, inner.Step.Name())
+				}
+			})
+		}
+		result = append(result, r)
+	})
+
+	serialized, err := json.Marshal(result)
+	if err != nil {
+		return fmt.Errorf("failed to marshal: %w", err)
+	}
+
+	dest := filepath.Join(artifactsEnv, "graph.json")
+	if err := ioutil.WriteFile(dest, serialized, 0644); err != nil {
+		return fmt.Errorf("failed to write %s: %w", dest, err)
+	}
+
+	return nil
+}
+
+func satisfiedBy(requirement api.StepLink, step api.Step) bool {
+	for _, creates := range step.Creates() {
+		if requirement.SatisfiedBy(creates) {
+			return true
+		}
+	}
+	return false
+}
+
+func iterateAllEdges(nodes []*api.StepNode, alreadyIterated sets.String, f func(*api.StepNode)) {
+	for _, node := range nodes {
+		iterateAllEdges(node.Children, alreadyIterated, f)
+		if alreadyIterated.Has(node.Step.Name()) {
+			continue
+		}
```

Comment:
Is there a reason why the `alreadyIterated` guard is done only after processing the children? I would think we could expect all children to be already seen too, no?

Reply:
Its done after processing the children because along the children chain there could be the current element, hence we avoid processing it twice. 

It can not be done before processing the children, because if we have an element `e` that is both a direct child of the initial element `i` and a grandchild via `i -> other-child -> e`, that latter `e` won't have its own children processed. No idea how likely this case is, but I don't really know much about the tree we are building here so I tried to avoid making assumptions.

Reply:
Wait you are right, its a graph not a tree and the elements are unique per name, so if we reached ourselves via a child before, we can assume that our children are processed already since we process children first.